Feat: bitbucket app

x-stp opened the pull request:
fork off brandonjyan:bitbucketapppassword which lived here https://github.com/trufflesecurity/trufflehog/pull/1498

The previous complex logic has been replaced with a clean, single-pass approach using a unified credentialPatterns slice for simplicity and performance.

The entire detector has been brought up to current project standards, attempt w/ cleaner regex with named capture groups and improved resource handling.

xxlarge bench is 2x faster.

 ### Checklist:

* [ ✔️ ] Tests passing (`make test-community`)?
   
* [ ✔️ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/welcome/install/#local-installation))?

brandonjyan committed:
add bitbucket app password scanner

brandonjyan committed:
clean up regex and username pattern logic

x-stp committed:
(chore): been a while, shall we merge off main?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/trufflesecurity/trufflehog?pullRequest=4214) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/trufflesecurity/trufflehog?pullRequest=4214) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: brandonjyan<br/>:x: x-stp<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/trufflesecurity/trufflehog?pullRequest=4214) it.</sub>

x-stp committed:
feat: re-intro bitbucket in engine.

x-stp committed:
feat: add BitbucketAppPassword detector type to proto files

x-stp committed:
Merge branch 'main' into feat/bitbucket_round2

amanfcp committed:
Merge branch 'main' into feat/bitbucket_round2

amanfcp commented on pkg/detectors/bitbucketapppassword/bitbucketapppassword.go:
```diff
@@ -0,0 +1,102 @@
+				if err == nil {
+					defer res.Body.Close()
+					// Status 403 FORBIDDEN indicates a valid secret without valid scope
+					if res.StatusCode >= 200 && res.StatusCode < 300 || res.StatusCode == 403 {
+						s1.Verified = true
+					}
```
Instead of using a broad conditional to check status codes, consider switching to a switch statement that explicitly handles expected responses like 200 OK and 401 Unauthorized. This makes the code easier to read, and allows for structured handling of each response case. It also ensures that unexpected status codes are surfaced clearly, which helps with debugging and observability.

Here's an [example](https://github.com/trufflesecurity/trufflehog/blob/f649faafce8c511ca0d53535e1387a3c42d4653b/pkg/detectors/ngrok/ngrok.go#L97)

x-stp replied:
Swapped. Did keep 200, 403 grouped in the switch block as they both return true.

x-stp committed:
Update pkg/detectors/bitbucketapppassword/bitbucketapppassword.go
Co-authored-by: Amaan Ullah <amaanuj.dev@gmail.com>

amanfcp commented on pkg/detectors/bitbucketapppassword/bitbucketapppassword.go:
```diff
@@ -0,0 +1,102 @@
+			}
+
+			if verify {
+				req, err := http.NewRequestWithContext(ctx, "GET", "https://api.bitbucket.org/2.0/user", nil)
```
```suggestion
				req, err := http.NewRequestWithContext(ctx, http.MethodGet, "https://api.bitbucket.org/2.0/user", nil)
```

amanfcp commented on pkg/detectors/bitbucketapppassword/bitbucketapppassword.go:
```diff
@@ -0,0 +1,102 @@
+				Raw:          []byte(fmt.Sprintf(`%s: %s`, resUsernameMatch, resAppPasswordMatch)),
+			}
+
+			if verify {
```
nit; In our new detectors, we do verification separately, see [this](https://github.com/trufflesecurity/trufflehog/blob/f649faafce8c511ca0d53535e1387a3c42d4653b/pkg/detectors/ngrok/ngrok.go#L62-L72)

amanfcp commented on pkg/detectors/bitbucketapppassword/bitbucketapppassword_test.go:
This should be renamed to bitbucketapppassword_integration_test.go.

amanfcp reviewed (commented):
We should also have a `bitbucketapppassword_test.go` for pattern test. See [this](https://github.com/trufflesecurity/trufflehog/blob/main/pkg/detectors/ngrok/ngrok_test.go)

x-stp commented on pkg/detectors/bitbucketapppassword/bitbucketapppassword.go:
```diff
@@ -0,0 +1,121 @@
+	for _, pattern := range credentialPatterns {
+		for _, match := range pattern.FindAllStringSubmatch(dataStr, -1) {
+			if len(match) == 3 {
+				username := strings.TrimSpace(match[1])
```
L63, L64: TrimSpace when the regex does not allow spaces %pval%p

x-stp committed:
refactor(bitbucket): tests (+patterns) ; code cleanup
- add switch{} block + credentialPattern[] for readability
- + tidy nested loops
- reflect codebase conventions in verification scope
- hard drain io.Discard, body.Close on res call in verify func
- intro the patterns test as it was not introoed
- standardize the integration test based off others living in repo

x-stp commented:
ran some benchmarks
where OLD == https://github.com/trufflesecurity/trufflehog/pull/1498


| Benchmark Size | GCP (Fastest) | Bitfinex | Hugging Face | Bitbucket NEW| Bitbucket OLD |
| :--- | :--- | :--- | :--- | :--- | :--- |
| **xsmall** | 585.6 | 1,156 | 599.2 | 923.8 | **507.1** |
| **small** | **681.6** | 1,435 | 986.4 | 6,125 | 10,420 |
| **medium** | **1,280** | 3,721 | 4,845 | 67,149 | 129,749 |
| **large** | **6,524** | 24,868 | 42,898 | 777,381 | 1,525,966 |
| **xlarge** | **60,241** | 240,401 | 414,083 | 7,247,065 | 17,305,812 |
| **xxlarge** | **748,382** | 2,563,427 | 4,265,669 | **75,731,043** | 145,927,025 |

still not really satisfied as these regexes are quite slow.

x-stp committed:
Merge branch 'main' into feat/bitbucket_round2

kashifkhan0771 committed:
Merge branch 'main' into feat/bitbucket_round2

amanfcp commented:
Thanks @x-stp for incorporating the pointers. Can you please resolve the conflicts?

x-stp committed:
Merge branch 'main' into feat/bitbucket_round2

amanfcp committed:
Merge branch 'main' into feat/bitbucket_round2

x-stp committed:
Merge branch 'main' into feat/bitbucket_round2

x-stp committed:
Update pkg/detectors/bitbucketapppassword/bitbucketapppassword.go

Co-authored-by: Amaan Ullah <amaanuj.dev@gmail.com>

kashifkhan0771 committed:
Merge branch 'main' into feat/bitbucket_round2

bugbaba commented on pkg/detectors/bitbucketapppassword/bitbucketapppassword.go:
```diff
@@ -0,0 +1,131 @@
+
+	var results []detectors.Result
+	for username, password := range uniqueCredentials {
+		result := detectors.Result{
```
Can we please use the standard `s1` variable name like the rest of detectors

bugbaba commented on pkg/detectors/bitbucketapppassword/bitbucketapppassword.go:
```diff
@@ -0,0 +1,131 @@
+	for username, password := range uniqueCredentials {
+		result := detectors.Result{
+			DetectorType: detectorspb.DetectorType_BitbucketAppPassword,
+			Raw:          fmt.Appendf(nil, "%s:%s", username, password),
```
Again please use the standard format

bugbaba commented on pkg/detectors/bitbucketapppassword/bitbucketapppassword.go:
```diff
@@ -0,0 +1,131 @@
+			if client == nil {
+				client = defaultClient
+			}
+			var vErr error
```
Why is

bugbaba commented on pkg/detectors/bitbucketapppassword/bitbucketapppassword.go:
```diff
@@ -0,0 +1,131 @@
+		}
+	}
+
+	var results []detectors.Result
```
Its better we stick to the format for this as well

bugbaba commented:
detector code already lacks uniformity across various files, we should try to stick to the core format as much as we can.

bugbaba commented on pkg/detectors/bitbucketapppassword/bitbucketapppassword.go:
```diff
@@ -0,0 +1,131 @@
+		return false, err
+	}
+	req.Header.Add("Accept", "application/json")
+	auth := base64.StdEncoding.EncodeToString(fmt.Appendf(nil, "%s:%s", username, password))
```
`SetBasicAuth()` should be used

bugbaba commented on pkg/detectors/bitbucketapppassword/bitbucketapppassword.go:
```diff
@@ -0,0 +1,131 @@
+
+	switch res.StatusCode {
+	case http.StatusOK, http.StatusForbidden:
+		// A 403 can indicate a valid credential with insufficient scope, which is still a finding.
```
Does it return any distinct message in response that we can check for?

x-stp commented:
> detector code already lacks uniformity across various files, we should try to stick to the core format as much as we can. 

Hi @bugbaba 

I will go over the comments soon and get back to you.

Thanks!
